lib/vm.py

```diff
@@ -361,6 +365,24 @@ def test_snapshot_on_running_vm(self):
         finally:
             snapshot.destroy(verify=True)
 
+    def vdi_resize(self, vdi_uuid, new_size=None):
+        # Reference
+        # MAX_VHD_SIZE = 2088960 * 1024 * 1024 # 2088960 MB
+        # MIN_VHD_SIZE = 1 * 1024 * 1024 # 1MB
+
+        # Following assert may be removed in future if we support online vdi resize
+        assert not self.is_running(), f"Expected halted, got running"
+
+        vdi_obj = VDI(vdi_uuid, sr=self.get_sr())
+        if new_size is None:
+            # Resize by double, ideally it should be within min and max limits reference
+            vdi_size = int(vdi_obj.param_get("virtual-size"))
+            new_size = str(vdi_size * 2)
```

Comment:
It would be cleaner to use integers to manipulate size objects.  `Host.xe()` already handles the conversion to string anyway.
To make such things more visible, it would be great to add type hints to new methods, eg:
```python
def vdi_resize(self, vdi_uuid: str, new_size: Optional[int] = None) -> None:
    ...
```